ML600 Update from bv platform - clean history

samuelvitorsaraiva opened the pull request.

magz97 committed:
fixed valve logic

magz97 committed:
updated to main

magz97 committed:
This makes it able to handle positions with None values as in the hamilton valves.

magz97 committed:
Fixed syntax and comment issue

magz97 committed:
Fixed syntax and comment issue

magz97 committed:
Fixed syntax and comment issue

samuelvitorsaraiva committed:
Ruff and mypy request to avoid mistakes!

JB-Wolf commented on src/flowchem/devices/hamilton/ml600.py:
```diff
@@ -91,9 +161,9 @@ def from_config(cls, config):
     async def initialize(self, hw_initialization: bool = True):
         """Ensure connection with pump and initialize it (if hw_initialization is True)."""
         self.num_pump_connected = await self._assign_pump_address()
```
This is misleading - the argument dioesnt do anything, the docstring is wrong. This causes unexpected bahaviour, please fix

samuelvitorsaraiva replied:
@magz97, can you address this?

magz97 committed:
Update ml600.py

Put back the hw initialization. This was taken out to avoid homing syringes if not wanted.

magz97 committed:
Update ml600.py

Initialize all hw instead of pump and valve separately

samuelvitorsaraiva committed:
update for ml600 from latest flowchem branch from the sugar platform.

samuelvitorsaraiva committed:
mypy and ruff checks

samuelvitorsaraiva committed:
Merge branch 'main' into update-from-BV-platform

samuelvitorsaraiva commented on src/flowchem/devices/hamilton/ml600.py:
```diff
@@ -380,24 +376,23 @@ def from_config(cls, **config):
+        )
 
-    async def initialize(self, hw_init=False, init_speed: str = "200 sec / stroke"):
+    async def initialize(self, init_speed: str = "200 sec / stroke"):
```
init_speed argument removed as at this stage only io is initialized

samuelvitorsaraiva committed:
Ruff checks

JB-Wolf commented on src/flowchem/devices/hamilton/ml600.py:
```diff
@@ -217,13 +247,9 @@ def _parse_response(self, response: str) -> str:
-        assert status in (self.ACKNOWLEDGE, self.NEGATIVE_ACKNOWLEDGE)
         if status == self.NEGATIVE_ACKNOWLEDGE:
             logger.warning("Negative acknowledge received")
-            warnings.warn(
```
why do we wanna lose this?

JB-Wolf commented on src/flowchem/devices/hamilton/ml600.py:
```diff
@@ -491,77 +477,82 @@ def _seconds_per_stroke_to_flowrate(self, second_per_stroke) -> float:
         await self.pause(pump)
         await self.send_command_and_read_reply(
-            Protocol1Command(command="", target_component=pump, execution_command="V"),)
+            Protocol1Command(command=ML600Commands.EMPTY, target_component=pump, execution_command="V"),)
```
Is return true really needed?

JB-Wolf commented on src/flowchem/devices/hamilton/ml600.py:
```diff
@@ -526,7 +580,7 @@ async def stop(self, pump: str = "") -> bool:
         return True  # Todo: need?
 
-    async def get_pump_status(self, pump: str = "") -> bool:
+    async def get_pump_status(self, pump: str) -> bool:
```
I think I also meant to maybe specify which pump is adressed.
Also here, A clear and unambiguous function name would be helpful, eg: pump_busy or pump_idle, then a bool as return would be very easy to understand

JB-Wolf commented on src/flowchem/devices/hamilton/ml600.py:
```diff
@@ -618,15 +673,15 @@ async def version(self) -> str:
         return await self.send_command_and_read_reply(Protocol1Command(command=ML600Commands.FIRMWARE_VERSION))
 
-    async def get_valve_angle(self, valve_code: str = "") -> int:
+    async def get_valve_angle(self, valve_code: str) -> int:
```
give options for valve code, also, I think the empty string as a default could be meant for single syringe setup - would be nice to check and potentially adapt

JB-Wolf commented on src/flowchem/devices/hamilton/ml600.py:
```diff
@@ -618,15 +673,15 @@ async def version(self) -> str:
     async def set_valve_angle(self,
                               target_angle: int,
-                              valve_code: str = "",
+                              valve_code: str,
```
also here, which values can valvecode adapt?

JB-Wolf commented on src/flowchem/devices/hamilton/ml600.py:
```diff
@@ -639,7 +694,7 @@ async def set_valve_angle(self,
         return True
 
-    async def get_valve_position_by_name(self, valve: str = "") -> str:
+    async def get_valve_position_by_name(self, valve: str) -> str:
```
also here, which values can valvecode adapt?

JB-Wolf commented on src/flowchem/devices/hamilton/ml600.py:
```diff
@@ -650,8 +705,8 @@ async def get_valve_position_by_name(self, valve: str = "") -> str:
         self,
-        valve: str = "",
-        target_position: str = "",
+        valve: str,
```
also here, which values can valvecode adapt?

JB-Wolf commented on src/flowchem/devices/hamilton/ml600.py:
```diff
@@ -669,7 +724,7 @@ async def set_valve_position_by_name(
         # todo: it's will be good check only pump but not whole system
 
-    async def get_raw_position(self, target_component: str = "") -> str:
+    async def get_raw_position(self, target_component: str) -> str:
```
also here, which values can target_component adapt?

JB-Wolf commented on src/flowchem/devices/hamilton/ml600.py:
```diff
@@ -707,6 +762,49 @@ async def set_raw_position(
+
+        This function was created specifically for the platform,
+        ensuring both syringes operate in perfect synchrony. Valve angles must previously
+        be explicitly set to control flow direction on each side.
```
_This function was created specifically for the platform,
ensuring both syringes operate in perfect synchrony. Valve angles must previously
be explicitly set to control flow direction on each side._
Use concise language:
This function was created for synchronously filling two syringes. Prior to executing this command, the valve positions have to be set correctly.

JB-Wolf commented on src/flowchem/devices/hamilton/ml600.py:
```diff
@@ -707,6 +762,49 @@ async def set_raw_position(
+                optional_parameter="S",
+                command_value=str(position),
+                parameter_value=set_speed_left,
+                target_component="B"
```
this is what I mean, just provide B and C as options in the docstring throughout

JB-Wolf commented on src/flowchem/devices/hamilton/ml600_pump.py:
```diff
@@ -32,6 +32,11 @@ def __init__(self, name: str, hw_device: ML600, pump_code: str = "") -> None:
             "" for single syringe pump. B or C  for dual syringe pump.
         """
         super().__init__(name, hw_device)
+        self.add_api_route("/set_to_volume", self.set_to_volume, methods=["PUT"])
```
I find set to volume weird -> maybe rename to move_syringe_to_volume or fill_syringe_to_volume? Depending on what it exactly does
Otherwise, it sounds like you are setting the syringe size

JB-Wolf commented on src/flowchem/devices/hamilton/ml600_pump.py:
```diff
@@ -51,7 +56,8 @@ async def is_pumping(self) -> bool:
         false means pump is not moving and buffer is empty. """
         # true might mean pump is moving, buffer still contain command or both
-        return await self.hw_device.get_pump_status(self.pump_code)
+        id_idle = await self.hw_device.is_idle(self.pump_code)
```
here, we have is_idle, why was there before this 
def status(self) -> bool: ?

JB-Wolf replied:
also, why is it id_idle and not is_idle? makes it harder to read

JB-Wolf commented on src/flowchem/devices/hamilton/ml600_pump.py:
```diff
@@ -68,7 +74,7 @@ async def stop(self) -> bool:
         if not await self.hw_device.get_pump_status(self.pump_code):
```
I do not think this sleep is a good idea -could this be modified / issue created?

JB-Wolf commented on src/flowchem/devices/hamilton/ml600_pump.py:
```diff
@@ -163,3 +169,59 @@ async def withdraw(self, rate: str = "1 ml/min", volume: str | None = None) -> b
         return await self.hw_device.get_pump_status(self.pump_code)
+
+    async def set_to_volume(self, volume: str, rate: str = "1 ml/min") -> bool:
+        """
```
this is a good docstring. It is clear and easy to understand what happens

JB-Wolf commented on src/flowchem/devices/hamilton/ml600_pump.py:
```diff
@@ -163,3 +169,59 @@ async def withdraw(self, rate: str = "1 ml/min", volume: str | None = None) -> b
+        await self.hw_device.set_to_volume(volume, rate, self.pump_code)
+        return await self.hw_device.get_pump_status(self.pump_code)
+
+    async def get_current_volume(self) -> float:
```
please checl the whole file, this function for example is declared twice from what I see. This must not happen!

JB-Wolf commented on src/flowchem/devices/hamilton/ml600.py:
```diff
@@ -446,7 +447,6 @@ async def set_return_steps(self, steps: int):
     async def initialize(self):
         """Initialize pump and its components."""
         await self.pump_io.initialize()
-        await self.wait_until_idle()
```
are you sure the wait until idle can be left out?

JB-Wolf commented on src/flowchem/devices/hamilton/ml600.py:
```diff
@@ -6,11 +6,9 @@
-from enum import Enum
 
 from flowchem import ureg
 from pint import Quantity
```
whats here? please fix this:
from flowchem import ureg
from pint import Quantity
Here, we import a slightly modified unit registry (ureg)
And the native one. Please make 100% sure that only the ureg, and NOT the native one are used
